runtime/centrifuge/src/migrations.rs

```diff
@@ -9,5 +9,92 @@
 
-pub type UpgradeCentrifuge1021 = ();
+use crate::{Investments, PoolSystem, Runtime, Weight};
+
+pub type UpgradeCentrifuge1021 = anemoy_pool::Migration;
+
+/// Migrate the Anemoy Pool's currency from LpEthUSC to Circle's USDC,
+/// native on Polkadot's AssetHub.
+mod anemoy_pool {
+
+	use cfg_primitives::{PoolId, TrancheId};
+	use cfg_types::tokens::{CurrencyId, TrancheCurrency};
+	use frame_support::traits::OnRuntimeUpgrade;
+	use pallet_pool_system::PoolDetailsOf;
+	#[cfg(feature = "try-runtime")]
+	use sp_std::vec::Vec;
+	#[cfg(feature = "try-runtime")]
+	use frame_support::ensure;
+	use cfg_types::orders::TotalOrder;
+
+	use super::*;
+
+	const ANEMOY_POOL_ID: PoolId = 4_139_607_887;
+	const LP_ETH_USDC: CurrencyId = CurrencyId::ForeignAsset(100_001);
+	const DOT_NATIVE_USDC: CurrencyId = CurrencyId::ForeignAsset(6);
+
+	pub struct Migration;
+
+	impl OnRuntimeUpgrade for Migration {
+		#[cfg(feature = "try-runtime")]
+		fn pre_upgrade() -> Result<Vec<u8>, &'static str> {
+			let pool_details: PoolDetailsOf<Runtime> =
+				PoolSystem::pool(ANEMOY_POOL_ID).ok_or("Could not find Anemoy Pool")?;
+
+			ensure!(
+				pool_details.currency == LP_ETH_USDC,
+				"anemoy_pool::Migration: pre_upgrade failing as Anemoy's currency should be LpEthUSDC"
+			);
+
+			Ok(pool_details.encode())
+		}
+
+		fn on_runtime_upgrade() -> Weight {
+			// To be executed at 1021, reject higher spec_versions
+			if crate::VERSION.spec_version >= 1022 {
+				log::info!("anemoy_pool::Migration: NOT execution since VERSION.spec_version >= 1022");
+				return Weight::zero();
+			}
+
+			pallet_pool_system::Pool::<Runtime>::mutate(ANEMOY_POOL_ID, |details| {
+				let details = details.as_mut().unwrap();
+				details.currency = DOT_NATIVE_USDC;
+				log::info!("anemoy_pool::Migration: Finished mutating currency to USDC");
+			});
+
+			<Runtime as frame_system::Config>::DbWeight::get().reads_writes(1, 1)
+		}
+
+		#[cfg(feature = "try-runtime")]
+		fn post_upgrade(old_state: Vec<u8>) -> Result<(), &'static str> {
+			let mut old_pool_details = PoolDetailsOf::<Runtime>::decode(&mut old_state.as_ref())
+				.map_err(|_| "Error decoding pre-upgrade state")?;
+
+			let pool_details: PoolDetailsOf<Runtime> =
+				PoolSystem::pool(ANEMOY_POOL_ID).ok_or("Could not find Anemoy Pool")?;
+
+			// Ensure the currency set to USDC is the only mutation performed
+			old_pool_details.currency = DOT_NATIVE_USDC;
+			ensure!(
+				old_pool_details == pool_details,
+				"Corrupted migration: Only the currency of the Anemoy pool should have changed"
+			);
+
+			log::info!("anemoy_pool::Migration: post_upgrade succeeded");
+			Ok(())
+		}
+	}
+
+	// todo(nuno): also check that pool value is 0 and check also that Investments::InvestOrders and
+	// Investments::RedeemOrder have no entries from Anemoy; the latter ones seem tricky at first
+	// sight since they are double maps first keyed by an AccountId, meaning we need to transverse
+	// that first which is more costly.
```

Comment:
Should suffice to check `StorageDoubleMap::<T>::iter_keys().count().is_zero()`.

Reply:
Can you expand on that? Because we need to iter keys on the account (first storage key) and then filter by the investment id (pool and tranche); if we just do iter_keys we are of taking any other pool into account.

If the idea is to use the assumption that anemoy is the only pool, then it makes sense. But it sort of defeats or weakens the point of this sanity check in the first place (tho, sure, it's less likely there will be another pool being added than a new order for anemoy).

Wtdyt? If everyone is fine with that I'm not blocking the approach

Reply:
I'll just filter the iter_keys involving anemoy to make this 100% robust.